### ISSUE TITLE
Enhancement: expose a dbus-type to plain english method

### ISSUE THREAD

**Issue by MathieuDuponchelle** [pwithnall/dbus-deviation#2]
Example : 'as' -> 'An array of strings'

Shouldn't be too tough, this is trivial to do for C when a decent AST is available for example. Is there any existing parsing mechanism of these types in dbus-deviation itself ?

**Comment by pwithnall** [comment]
> Example : 'as' -> 'An array of strings'

Great idea.

> Shouldn't be too tough, this is trivial to do for C when a decent AST is available for example. Is there any existing parsing mechanism of these types in dbus-deviation itself ?

Potential complication: this needs to be localisable, so we basically need to write a separate implementation for each language which dbus-deviation is going to be used for. For the moment, that could just be English, but I would want the interface to have the potential to support other languages in future.

There is no existing parsing mechanism. We could use `GVariantType` (it’s a superset of the D-Bus type system — see [the documentation](https://developer.gnome.org/glib/stable/glib-GVariantType.html#glib-GVariantType.description)), but that would mean adding a dependency on GLib. Not so bad.

[dbus-python implements the D-Bus type system](http://dbus.freedesktop.org/doc/dbus-python/doc/tutorial.html#data-types), however:
 * dbus-python is old and we don’t want to encourage its use
 * It tries to merge the D-Bus and Python type systems in a way which always ends up being a bit ugly because it tries to make too many things implicit rather than explicit; let’s not encourage that

Since we’re only dealing with the type system, not runtime instances of any of those types, I’d suggest either using `GVariantType` or implementing something similar to it.

Patches welcome.

**Comment by zayfod** [comment]
The type parser that I have been working on seems to be a step in the necessary direction for this feature. It would need some rework after the lxml integration but it is coming...

**Comment by MathieuDuponchelle** [comment]
Cool :) As for localization @pwithnall it might be tricky to get right, plain english would already be awesome for the generated documentation :)

**Comment by zayfod** [comment]
Here is what is possible with the new parser. Does it make sense to you? Any better suggestions?

"i": "32-bit integer"
"ii": "32-bit integer, 32-bit integer"
"aiai": "array of 32-bit integer, array of 32-bit integer"
"u": "unsigned 32-bit integer"
"uu": "unsigned 32-bit integer, unsigned 32-bit integer"
"ai": "array of 32-bit integer"
"aai": "array of array of 32-bit integer"
"au": "array of unsigned 32-bit integer"
"av": "array of variant"
"(iii)": "structure with 32-bit integer, 32-bit integer, 32-bit integer fields"
"(ii)(ii)": "structure with 32-bit integer, 32-bit integer fields, structure with 32-bit integer, 32-bit integer fields"
"(i(ii))": "structure with 32-bit integer, structure with 32-bit integer, 32-bit integer fields fields"
"(v)": "structure with variant fields"
"(ius)": "structure with 32-bit integer, unsigned 32-bit integer, string fields"
"a(ii)": "array of structure with 32-bit integer, 32-bit integer fields"
"a{us}": "array of dictionary mapping unsigned 32-bit integer to string"
"a{us}i": "array of dictionary mapping unsigned 32-bit integer to string, 32-bit integer"

**Comment by MathieuDuponchelle** [comment]
That looks great! Can't wait to integrate this in my documentation tool :)

**Comment by MathieuDuponchelle** [comment]
No specific suggestions by the way, I'm not a DBus expert by any means, I only can tell it would help me immensely for that exact reason to have the types for a DBus interface spelled out to me explicitly.

**Comment by zayfod** [comment]
OK, I'll add this right after #13 :)

**Comment by pwithnall** [comment]
> "(ii)(ii)": "structure with 32-bit integer, 32-bit integer fields, structure with 32-bit integer, 32-bit integer fields"

The English version of that has lost a lot of information compared to the encoded version. What I find hard about D-Bus types is remembering what each letter means (for example, what’s the letter to represent a signed 16-bit integer??). How about an English version which expands these letters, but continues to use the bracketing syntax to ensure the nesting of types is clear?

So for this example, that would be:
> (32-bit integer, 32-bit integer) (32-bit integer, 32-bit integer)

I would expect this to be implemented as a separate class (i.e. not part of the `TypeSignature` AST) — something like `TypeFormatter` which takes in a `TypeSignature`. That way, the AST classes don’t get clogged up with methods to format them in multiple different ways.

**Comment by zayfod** [comment]
My personal preference is to stick to the original goal of "plain English D-Bus types".

While I see the point and I see value for hard core D-Bus users, this approach is less helpful for casual users that will benefit the most from "plain English D-Bus types".

What do you think guys?

@pwithnall, in the end I'll follow your direction.

**Comment by pwithnall** [comment]
I’m happy with whatever, as long as it makes the structure of the type obvious; the suggestions from https://github.com/pwithnall/dbus-deviation/issues/2#issuecomment-193558607 don’t make this clear enough.

**Comment by MathieuDuponchelle** [comment]
@pwithnall for my use case I'd rather have the ultra hand-holdy version, FWIW hotdoc will display both prototypes. I agree structure might be made more obvious, but "structure with" (or "structure containing") should still be in there, as a DBus noob I didn't know there was such a concept as a "DBus structure". Just my 2 cents anyway :)

**Comment by pwithnall** [comment]
> as a DBus noob I didn't know there was such a concept as a "DBus structure". 

I don’t think it’s feasible to target documentation about D-Bus APIs to developers who don’t know the first thing about D-Bus. That would be like changing all the C documentation from saying ‘const char *’ to saying ‘pointer to a sequence of constant signed bytes’. There’s a balance between verbosity and helpfulness, and I think that in the general case, using words to describe everything in a D-Bus type is too verbose, as well as being very complex to implement satisfactorily, and even harder to translate (which needs to be supported).

What the documentation could do is put a small link to an introduction to the D-Bus type system next to each type, so that newbies do have //somewhere// to go to learn about the types.

**Comment by MathieuDuponchelle** [comment]
Well I suppose we could provide various levels of verbosity ? Could be reflected in hotdoc's interface as well, with the maximum level visible through a "show-more" button or somesuch. My initial request was for plain english types, and there was a reason for that, I understand your use case is different but there's no reason it should be the only valid one ;)

**Comment by pwithnall** [comment]
I don’t think plain English types can work. The complexity of producing unambiguous text which describes a complex D-Bus type, but which is also readable and, importantly, understandable by someone who doesn’t know D-Bus at all, is pretty tough.

What is your reason for wanting plain English types?

One approach you could take would be to draw the type AST in the documentation, e.g. for the `(ii)(ii)` type above:
```
Structure
 |- 32-bit integer
 \- 32-bit integer
Structure
 |- 32-bit integer
 \- 32-bit integer
```

That avoids the English-language complexity, while also avoiding D-Bus bracketing syntax. But really, someone who doesn’t know the basics of the D-Bus type system is not going to get very far with using a D-Bus API, no matter how verbose the API documentation is, unless they read up on the type system.

**Comment by MathieuDuponchelle** [comment]
I don't think it's that tough, I remember writing something similar to [this](http://cdecl.org/) in an introductory course to C AST manipulation. surely the complexity is similar. The reason why I want this is because I find these prototypes a bit cryptic, and I think deciphering them can be a valuable feature for a documentation tool, not super critical but nice to have :)

**Comment by pwithnall** [comment]
PR #16 created to implement this. Feedback on the format welcome before I merge it.

I haven’t implemented translation support (because the rest of dbus-deviation doesn’t have it), but it can be added in future as a property of `BasicTypeFormatter`.

**Comment by MathieuDuponchelle** [comment]
\o/ , I will use this in hotdoc as soon as there's a version on pypi , no particular comment on the pull request, seems appropriate and well-tested.

**Comment by pwithnall** [comment]
Done. I’ll make a release soon.
